src/mcts/search.h

```diff
@@ -282,50 +301,143 @@ class SearchWorker {
+        const auto& move = moves[last_idx++];
+        if (last_idx == moves.size()) last_idx = 0;
+        if (move.first == move_id) return move.second;
+        ++total_count;
+      }
+      assert(false);  // Move not found.
+      return 0;
     }
 
    private:
-    NodeToProcess(Node* node, uint16_t depth, bool is_collision, int multivisit)
+    NodeToProcess(Node* node, uint16_t depth, bool is_collision, int multivisit,
+                  int maxvisit)
         : node(node),
           multivisit(multivisit),
+          maxvisit(maxvisit),
           depth(depth),
           is_collision(is_collision) {}
   };
+  // Holds per task worker scratch data
+  struct TaskWorkspace {
+    Node::Iterator cur_iters[256];
+    std::vector<std::unique_ptr<int[]>> vtp_buffer;
+  };
 
-  NodeToProcess PickNodeToExtend(int collision_limit);
-  void ExtendNode(Node* node, int depth);
-  bool AddNodeToComputation(Node* node, bool add_if_cached, int* transform_out);
+  void PickNodesToExtend(int collision_limit);
+  void PickNodesToExtendTask(Node* starting_point, int collision_limit,
+                             int base_depth,
+                             const std::vector<Move>& moves_to_base,
+                             std::vector<NodeToProcess>* receiver,
+                             TaskWorkspace* workspace);
+  void ProcessPickedTask(int batch_start, int batch_end,
+                         TaskWorkspace* workspace);
+  void ExtendNode(Node* node, int depth, const std::vector<Move>& moves_to_add,
+                  PositionHistory* history);
+  bool AddNodeToComputation(Node* node, bool add_if_cached, int* transform_out,
+                            PositionHistory* history);
   int PrefetchIntoCache(Node* node, int budget, bool is_odd_depth);
+  template <typename Computation>
   void FetchSingleNodeResult(NodeToProcess* node_to_process,
+                             const Computation& computation,
                              int idx_in_computation);
   void DoBackupUpdateSingleNode(const NodeToProcess& node_to_process);
   // Returns whether a node's bounds were set based on its children.
-  bool MaybeSetBounds(Node* p, float m, int* n_to_fix, float* v_delta, float* d_delta, float* m_delta) const;
+  bool MaybeSetBounds(Node* p, float m, int* n_to_fix, float* v_delta,
+                      float* d_delta, float* m_delta) const;
+
+  void RunTasks(int tid);
 
   Search* const search_;
   // List of nodes to process.
   std::vector<NodeToProcess> minibatch_;
   std::unique_ptr<CachingComputation> computation_;
-  // History is reset and extended by PickNodeToExtend().
+  // History is reset and extended by PrefetchIntoCache().
   PositionHistory history_;
   int number_out_of_order_ = 0;
   const SearchParams& params_;
   std::unique_ptr<Node> precached_node_;
   const bool moves_left_support_;
   IterationStats iteration_stats_;
   StoppersHints latest_time_manager_hints_;
+
+  struct PickTask {
```

Comment:
That's quite a weird struct, constructor declared in the middle of variables, half of variables are with trailing underscores while the other half is not.

Ideally it would be nice to have it converted into a class (with all member variables private), but it's a good idea at least to tidy it up.

Reply:
I'm thinking you may have misread this code a bit.  The constructors are at the end, there are no variables after the constructors.  What I will take as a point I had intended to cleanup was that I should move it as a whole up to be next to NodeToProcess rather than being here in the middle of the SearchWorker field declarations.